Feat: Add KMS methods for Key Rotations and MACs

MauriceBrg opened the pull request:
Implemented New Methods for KMS:

- `generate_mac`
- `rotate_key_on_demand`
- `list_key_rotations`
- `verify_mac`

Added these already implemented methods as dummies to the Backend so they show up in the coverage:

- `list_key_policies`
- `generate_random`

(When I ran `scripts/implementation_coverage.py` it also updated a bunch of other rst-files (>50) with new methods, I'm not sure if I should commit those as well, as most are unrelated to KMS)

MauriceBrg commented:
The two lines missing coverage are for the dummy methods to include `list_key_policies` and `generate_random` in the implementation coverage. Should I write some dummy tests or call them from the responses?

bblommers commented on moto/kms/models.py:
```diff
@@ -714,5 +731,84 @@ def get_public_key(self, key_id: str) -> Tuple[Key, bytes]:
+        key: Key = self.keys[self.get_key_id(key_id)]
+
+        # we treat the marker as an index-offset
+        marker = marker or "0"
```
Does AWS use integers as well for markers, or do they use a random string? If it's random, we can use our '@paginate`-marker - that should simplify the implementation a lot.

See the docs here: https://docs.getmoto.org/en/latest/docs/contributing/development_tips/utilities.html#paginator

MauriceBrg replied:
Somehow missed that this exists, but I should have known that I'm not the first to build a pagination scheme.
I have changed the code to use that utility.

bblommers commented on moto/kms/models.py:
```diff
@@ -714,5 +731,84 @@ def get_public_key(self, key_id: str) -> Tuple[Key, bytes]:
+        if mac != regenerated_mac:
+            raise KMSInvalidMacException()
+
+        return key_id, True, mac_algorithm
```
Do we need to return anything here? We already know the `key_id` and `mac_algorithm`, and the `is_valid` will always be `True` - so we do not return any new information. Not returning anything would simplify this method

MauriceBrg replied:
Makes sense, I shouldn't have blindly followed the scaffold-layout :)

I'll update it.

bblommers commented:
Hi @MauriceBrg, thanks for the PR! Just left two comments.

> (When I ran scripts/implementation_coverage.py it also updated a bunch of other rst-files (>50) with new methods, I'm not sure if I should commit those as well, as most are unrelated to KMS)

That will be either because Moto hasn't been updated in a while, or you're running the script with an outdated version of `botocore`. Either way, it's not necessary - the implementation docs will always be updated when we release a new version.

> The two lines missing coverage are for the dummy methods to include list_key_policies and generate_random in the implementation coverage. Should I write some dummy tests or call them from the responses?

I'm OK with the current approach. Having 100% coverage is a nice to have, but in the end, it doesn't functionally change anything. We have an implementation, and we have tests for it - that's all I care about.

MauriceBrg commented:
Hey @bblommers 

thanks for the suggestions, I have updated the code to address them and removed the update `IMPLEMENTATION_COVERAGE.md` for this Pull Request :)

I've updated boto3 and botocore, now the `scripts/implementation_coverage.py` would only update a handful of files, but I'll leave that to you for the next release.

```text
boto3==1.35.92
botocore==1.35.92
```

bblommers reviewed (approved):
Great, thank you @MauriceBrg!
